File: iteration2/src/course/TechnicalElectiveCourse.java
package iteration2.src.course;

import iteration2.src.Department;
import iteration2.src.human.Assistant;
import iteration2.src.human.Grade;
import iteration2.src.human.Lecturer;
import iteration2.src.human.Student;
import iteration2.src.input_output.Logger;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TechnicalElectiveCourse extends ElectiveCourse{
    public static final int REQUIRED_CREDITS = 155;

    public static final Map<Integer, Integer> numberOfCoursesTakeableBySemester = new HashMap<>(){{
        put(0,0);
        put(1,0);
        put(2,0);
        put(3,0);
        put(4,0);
        put(5,0);
        put(6,1);
        put(7, 3);
    }};

    public TechnicalElectiveCourse(String code, String name, int credits, int theoreticalHours, int appliedHours,
                                   Grade firstYearToTake, Season firstSeasonToTake, List<Lecturer> lecturers, List<Assistant> assistants){
        super(code,name,credits,theoreticalHours,appliedHours,firstYearToTake,firstSeasonToTake,lecturers,assistants);
    }

    @Override
    public void addPrerequisite(Course prerequisite){
        prerequisites.add(prerequisite);
    }

    @Override
    public Boolean canStudentTakeCourse(Student student) {
        int semester = student.getStudentSemester();
        int noOfCoursesTillSemester = getTotalNumberOfCoursesUntilSemester(semester);

        return student.getTranscript().getNumberOfTElectivesPassed() < noOfCoursesTillSemester
        && !student.didStudentPass(this)
        && isCreditsRequirementMet(student);
    }

    public boolean isCreditsRequirementMet(Student student){
        return REQUIRED_CREDITS <= student.getCompletedCredits();
    }

    public static int getTotalNumberOfCoursesUntilSemester(int semester){
        int noOfCoursesTillSemester = 0;

        for(int i = 0; i <= semester; i++){
            noOfCoursesTillSemester += numberOfCoursesTakeableBySemester.get(i).intValue();
        }

        return noOfCoursesTillSemester;
    }

    public int getCoursePriority(){
        return 2;
    }
}

File: iteration3/TechnicalElectiveCourse.py
class TechnicalElectiveCourse(ElectiveCourse):
    REQUIRED_CREDITS = 155

    numberOfCoursesTakeableBySemester = {
        0: 0,
        1: 0,
        2: 0,
        3: 0,
        4: 0,
        5: 0,
        6: 1,
        7: 3
    }

    def __init__(self, code, name, credits, theoretical_hours, applied_hours, first_year_to_take, first_season_to_take, lecturers, assistants):
        super().__init__(code, name, credits, theoretical_hours, applied_hours, first_year_to_take, first_season_to_take, lecturers, assistants)

    def add_prerequisite(self, prerequisite):
        self.prerequisites.add(prerequisite)

    def can_student_take_course(self, student):
        semester = student.get_student_semester()
        no_of_courses_till_semester = self.get_total_number_of_courses_until_semester(semester)

        return student.get_transcript().get_number_of_te_courses_passed() < no_of_courses_till_semester and not student.did_student_pass(self) and self.is_credits_requirement_met(student)

    def is_credits_requirement_met(self, student):
        return self.REQUIRED_CREDITS <= student.get_completed_credits()

    @staticmethod
    def get_total_number_of_courses_until_semester(semester):
        no_of_courses_till_semester = 0

        for i in range(semester + 1):
            no_of_courses_till_semester += TechnicalElectiveCourse.numberOfCoursesTakeableBySemester[i]

        return no_of_courses_till_semester

    def get_course_priority(self):
        return 2
